File: api/src/main/java/org/pmiops/workbench/db/model/CohortReview.java
package org.pmiops.workbench.db.model;

import java.sql.Timestamp;
import java.util.Objects;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Lob;
import javax.persistence.Table;
import javax.persistence.Transient;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.pmiops.workbench.model.ReviewStatus;

@Entity
@Table(name = "cohort_review")
public class CohortReview {

    private long cohortReviewId;
    private long cohortId;
    private long cdrVersionId;
    private Timestamp creationTime;
    private String cohortDefinition;
    private String cohortName;
    private Timestamp lastModifiedTime;
    private long matchedParticipantCount;
    private long reviewSize;
    private long reviewedCount;
    private Short reviewStatus;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "cohort_review_id")
    public long getCohortReviewId() {
        return cohortReviewId;
    }

    public void setCohortReviewId(long cohortReviewId) {
        this.cohortReviewId = cohortReviewId;
    }

    public CohortReview cohortReviewId(long cohortReviewId) {
        this.cohortReviewId = cohortReviewId;
        return this;
    }

    @Column(name = "cohort_id")
    public long getCohortId() {
        return cohortId;
    }

    public void setCohortId(long cohortId) {
        this.cohortId = cohortId;
    }

    public CohortReview cohortId(long cohortId) {
        this.cohortId = cohortId;
        return this;
    }

    @Column(name = "cdr_version_id")
    public long getCdrVersionId() {
        return cdrVersionId;
    }

    public void setCdrVersionId(long cdrVersionId) {
        this.cdrVersionId = cdrVersionId;
    }

    public CohortReview cdrVersionId(long cdrVersionId) {
        this.cdrVersionId = cdrVersionId;
        return this;
    }

    @Column(name = "cohort_name")
    public String getCohortName() {
        return cohortName;
    }

    public void setCohortName(String cohortName) {
        this.cohortName = cohortName;
    }

    public CohortReview cohortName(String cohortName) {
        this.cohortName = cohortName;
        return this;
    }

    @Lob
    @Column(name = "cohort_definition")
    public String getCohortDefinition() {
        return cohortDefinition;
    }

    public void setCohortDefinition(String cohortDefinition) {
        this.cohortDefinition = cohortDefinition;
    }

    public CohortReview cohortDefinition(String cohortDefinition) {
        this.cohortDefinition = cohortDefinition;
        return this;
    }

    @Column(name = "creation_time")
    public Timestamp getCreationTime() {
        return creationTime;
    }

    public void setCreationTime(Timestamp creationTime) {
        this.creationTime = creationTime;
    }

    public CohortReview creationTime(Timestamp creationTime) {
        this.creationTime = creationTime;
        return this;
    }

    @Column(name = "last_modified_time")
    public Timestamp getLastModifiedTime() {
        return lastModifiedTime;
    }

    public void setLastModifiedTime(Timestamp lastModifiedTime) {
        this.lastModifiedTime = lastModifiedTime;
    }

    public CohortReview lastModifiedTime(Timestamp lastModifiedTime) {
        this.lastModifiedTime = lastModifiedTime;
        return this;
    }

    @Column(name = "matched_participant_count")
    public long getMatchedParticipantCount() {
        return matchedParticipantCount;
    }

    public void setMatchedParticipantCount(long matchedParticipantCount) {
        this.matchedParticipantCount = matchedParticipantCount;
    }

    public CohortReview matchedParticipantCount(long matchedParticipantCount) {
        this.matchedParticipantCount = matchedParticipantCount;
        return this;
    }

    @Column(name = "review_size")
    public long getReviewSize() {
        return reviewSize;
    }

    public void setReviewSize(long reviewSize) {
        this.reviewSize = reviewSize;
    }

    public CohortReview reviewSize(long reviewSize) {
        this.reviewSize = reviewSize;
        return this;
    }

    public void incrementReviewedCount() {
        this.reviewedCount = reviewedCount + 1;
    }

    @Column(name = "reviewed_count")
    public long getReviewedCount() {
        return reviewedCount;
    }

    public void setReviewedCount(long reviewedCount) {
        this.reviewedCount = reviewedCount;
    }

    public CohortReview reviewedCount(long reviewedCount) {
        this.reviewedCount = reviewedCount;
        return this;
    }

    @Column(name = "review_status")
    public Short getReviewStatus() {
        return reviewStatus;
    }

    public void setReviewStatus(Short reviewStatus) {
        this.reviewStatus = reviewStatus;
    }

    public CohortReview reviewStatus(Short reviewStatus) {
        this.reviewStatus = reviewStatus;
        return this;
    }

    @Transient
    public ReviewStatus getReviewStatusEnum() {
        return StorageEnums.reviewStatusFromStorage(getReviewStatus());
    }

    public void setReviewStatusEnum(ReviewStatus reviewStatus) {
        setReviewStatus(StorageEnums.reviewStatusToStorage(reviewStatus));
    }

    public CohortReview reviewStatusEnum(ReviewStatus reviewStatus) {
        return this.reviewStatus(StorageEnums.reviewStatusToStorage(reviewStatus));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CohortReview that = (CohortReview) o;
        return cohortId == that.cohortId &&
                cdrVersionId == that.cdrVersionId &&
                matchedParticipantCount == that.matchedParticipantCount &&
                reviewSize == that.reviewSize &&
                reviewedCount == that.reviewedCount &&
                Objects.equals(lastModifiedTime, that.lastModifiedTime) &&
                reviewStatus == that.reviewStatus;
    }

    @Override
    public int hashCode() {
        return Objects.hash(cdrVersionId, lastModifiedTime, matchedParticipantCount, reviewSize, reviewedCount, reviewStatus);
    }

    @Override
    public String toString() {
        return new ToStringBuilder(this)
                .append("cohortReviewId", cohortReviewId)
                .append("cohortId", cohortId)
                .append("cdrVersionId", cdrVersionId)
                .append("creationTime", creationTime)
                .append("lastModifiedTime", lastModifiedTime)
                .append("matchedParticipantCount", matchedParticipantCount)
                .append("reviewSize", reviewSize)
                .append("reviewedCount", reviewedCount)
                .append("reviewStatus", reviewStatus)
                .toString();
    }
}

File: api/src/main/java/org/pmiops/workbench/db/model/CohortReview.kt
package org.pmiops.workbench.db.model

import java.sql.Timestamp
import java.util.Objects
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.Lob
import javax.persistence.ManyToOne
import javax.persistence.Table
import javax.persistence.Transient
import javax.persistence.Version
import org.apache.commons.lang3.builder.ToStringBuilder
import org.pmiops.workbench.model.ReviewStatus

@Entity
@Table(name = "cohort_review")
class CohortReview {

    @get:Id
    @get:GeneratedValue(strategy = GenerationType.IDENTITY)
    @get:Column(name = "cohort_review_id")
    var cohortReviewId: Long = 0
    @get:Version
    @get:Column(name = "version")
    var version: Int = 0
    @get:Column(name = "cohort_id")
    var cohortId: Long = 0
    @get:Column(name = "cdr_version_id")
    var cdrVersionId: Long = 0
    @get:Column(name = "creation_time")
    var creationTime: Timestamp? = null
    @get:Lob
    @get:Column(name = "cohort_definition")
    var cohortDefinition: String? = null
    @get:Column(name = "cohort_name")
    var cohortName: String? = null
    @get:Column(name = "description")
    var description: String? = null
    @get:Column(name = "last_modified_time")
    var lastModifiedTime: Timestamp? = null
    @get:Column(name = "matched_participant_count")
    var matchedParticipantCount: Long = 0
    @get:Column(name = "review_size")
    var reviewSize: Long = 0
    @get:Column(name = "reviewed_count")
    var reviewedCount: Long = 0
    @get:Column(name = "review_status")
    var reviewStatus: Short? = null
    @get:ManyToOne
    @get:JoinColumn(name = "creator_id")
    var creator: User? = null

    var reviewStatusEnum: ReviewStatus
        @Transient
        get() = StorageEnums.reviewStatusFromStorage(reviewStatus)
        set(reviewStatus) {
            reviewStatus = StorageEnums.reviewStatusToStorage(reviewStatus)
        }

    fun cohortReviewId(cohortReviewId: Long): CohortReview {
        this.cohortReviewId = cohortReviewId
        return this
    }

    fun version(version: Int): CohortReview {
        this.version = version
        return this
    }

    fun cohortId(cohortId: Long): CohortReview {
        this.cohortId = cohortId
        return this
    }

    fun cdrVersionId(cdrVersionId: Long): CohortReview {
        this.cdrVersionId = cdrVersionId
        return this
    }

    fun cohortName(cohortName: String): CohortReview {
        this.cohortName = cohortName
        return this
    }

    fun description(description: String): CohortReview {
        this.description = description
        return this
    }

    fun cohortDefinition(cohortDefinition: String): CohortReview {
        this.cohortDefinition = cohortDefinition
        return this
    }

    fun creationTime(creationTime: Timestamp): CohortReview {
        this.creationTime = creationTime
        return this
    }

    fun lastModifiedTime(lastModifiedTime: Timestamp): CohortReview {
        this.lastModifiedTime = lastModifiedTime
        return this
    }

    fun matchedParticipantCount(matchedParticipantCount: Long): CohortReview {
        this.matchedParticipantCount = matchedParticipantCount
        return this
    }

    fun reviewSize(reviewSize: Long): CohortReview {
        this.reviewSize = reviewSize
        return this
    }

    fun incrementReviewedCount() {
        this.reviewedCount = reviewedCount + 1
    }

    fun reviewedCount(reviewedCount: Long): CohortReview {
        this.reviewedCount = reviewedCount
        return this
    }

    fun reviewStatus(reviewStatus: Short?): CohortReview {
        this.reviewStatus = reviewStatus
        return this
    }

    fun reviewStatusEnum(reviewStatus: ReviewStatus): CohortReview {
        return this.reviewStatus(StorageEnums.reviewStatusToStorage(reviewStatus))
    }

    fun creator(creator: User): CohortReview {
        this.creator = creator
        return this
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as CohortReview?
        return (version == that!!.version
                && cohortId == that.cohortId
                && cdrVersionId == that.cdrVersionId
                && matchedParticipantCount == that.matchedParticipantCount
                && reviewSize == that.reviewSize
                && reviewedCount == that.reviewedCount
                && creationTime == that.creationTime
                && cohortDefinition == that.cohortDefinition
                && cohortName == that.cohortName
                && description == that.description
                && lastModifiedTime == that.lastModifiedTime
                && reviewStatus == that.reviewStatus)
    }

    override fun hashCode(): Int {
        return Objects.hash(
                version,
                cohortId,
                cdrVersionId,
                creationTime,
                cohortDefinition,
                cohortName,
                description,
                lastModifiedTime,
                matchedParticipantCount,
                reviewSize,
                reviewedCount,
                reviewStatus)
    }

    override fun toString(): String {
        return ToStringBuilder(this)
                .append("cohortReviewId", cohortReviewId)
                .append("version", version)
                .append("cohortId", cohortId)
                .append("cdrVersionId", cdrVersionId)
                .append("creationTime", creationTime)
                .append("cohortDefinition", cohortDefinition)
                .append("cohortName", cohortName)
                .append("description", description)
                .append("lastModifiedTime", lastModifiedTime)
                .append("matchedParticipantCount", matchedParticipantCount)
                .append("reviewSize", reviewSize)
                .append("reviewedCount", reviewedCount)
                .append("reviewStatus", reviewStatus)
                .toString()
    }
}
